torch/_dynamo/variables/builder.py

```diff
@@ -474,6 +475,30 @@ def index_source(key):
                 source=self.source,
                 guards=make_guards(GuardBuilder.FUNCTION_MATCH),
             )
+        elif isinstance(value, functools.partial):
+            func_src = AttrSource(self.get_source(), "func")
+            func_obj = VariableBuilder(self.tx, func_src)(value.func)
+
+            args = []
+            args_source = AttrSource(self.get_source(), "args")
+            for i, arg in enumerate(value.args):
+                args.append(
+                    VariableBuilder(self.tx, GetItemSource(args_source, i))(arg)
+                )
+
+            keywords = {}
+            keywords_source = AttrSource(self.get_source(), "keywords")
+            for k, v in value.keywords.items():
+                keywords[k] = VariableBuilder(
+                    self.tx, GetItemSource(keywords_source, k)
+                )(v)
+
+            guards = {
+                keywords_source.make_guard(GuardBuilder.DICT_KEYS),
+                args_source.make_guard(GuardBuilder.LIST_LENGTH),
+            }
+
+            return FunctoolsPartialVariable(func_obj, args, keywords, guards=guards)
```

Comment:
We are missing a TYPE_MATCH guard.  We need to make sure `self.get_source()` is a `functools.partial` instance before blinding reading its attributes.

Please add a test where the type changes from functools.partial into something else on a second invocation.

You can also do:
`.add_options(func_obj, args, keywords)`

to copy the guards from the child objects rather than manually constructing duplicate guards above.

Reply:
SGTM added test. 

> You can also do:
> .add_options(func_obj, args, keywords)
> to copy the guards from the child objects rather than manually constructing duplicate guards above.


We can't only have the add_options here - the guard propagation must be in the ctor because you can make a partial under a compile region from an input.